### PR TITLE
feat: ✨ add sd-badge (#232)

### PR DESCRIPTION
**Pull request by MarcMatthiae**
## Description:
I added the sd-badge component including storybook stories and tests.

## Definition of Reviewable:
- [x] Documentation is created/updated
- [x] relevant tickets are linked
- [x] PR is assigned to project board

### CODE REVIEW

File: packages/components/src/components/badge/badge.stories.ts [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+const { generateTemplate } = storybookTemplate('sd-badge');
+
+export default {
+  title: 'Components/sd-badge',
+  component: 'sd-badge',
+  parameters: {
+    ...parameters,
+    design: {
+      type: 'figma',
+      url: 'https://www.figma.com/file/fPGhgNZv98U4H69Gu2tlWi/Button?type=design&node-id=13-18&t=jDLqFEdY7ZlOJurc-4'
```
Comment (karlbaumhauer) [inline packages/components/src/components/badge/badge.stories.ts:18]: still the Button URL

File: packages/components/src/components/badge/badge.stories.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+      args
+    });
+  }
+};
+
+/**
+ * Use the `default` slot to add content to the badge.
+ * Use the `overflow-indicator` slot to add an overflow indicator when the parameter <i>overflowing</i> is true.
+ *
+ * <b>Why did we add a slot for the overflow-indicator?</b>
```
Comment (karlbaumhauer) [inline packages/components/src/components/badge/badge.stories.ts:160]: seems like a dev-comment...?!? pls remove

File: packages/components/src/components/badge/badge.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import { HasSlotController } from '../../internal/slot';
+import { html } from 'lit';
+import cx from 'classix';
+import SolidElement from '../../internal/solid-element';
+
+/**
+ * @summary Badges are used to draw attention and display statuses or counts.
+ * @documentation https://solid.union-investment.com/[storybook-link]/badge
+ * @status stable
+ * @since 1.0
```
Comment (karlbaumhauer) [inline packages/components/src/components/badge/badge.ts:11]: shouldnt this be the future version when merged to `main`?

**Review by karlbaumhauer — CHANGES_REQUESTED** [review]

**Review by karlbaumhauer — APPROVED** [review]

**Comment by mariohamann** [comment]
CleanShot 2023-08-10 at 15 20 53@2x

The Overflow indicator is not vertically centered. And it needs a padding to the left.

CleanShot 2023-08-10 at 15 21 52@2x

1. This prop shouldn't exist, it should be handled via slot.
2. You shouldn't set the default in the story, otherwise it's ignored if you want to do something.
3. Add doumentation, that you should add the plus here to see it.

**Comment by mariohamann** [comment]
CleanShot 2023-08-10 at 15 23 58@2x

Please show this with md and sm buttons, reflecting the size of the badge as well.

**Comment by mariohamann** [comment]
CleanShot 2023-08-10 at 15 25 02@2x

Why doesn't the overflow indicator show up? Please only use the slot, no prop for it.

**Comment by mariohamann** [comment]
CleanShot 2023-08-10 at 15 28 32@2x

I'm not sure if this behaviour is correct. :)

I think it should be right-aligned, but maybe the designers have another opinion on that?

File: packages/components/src/components/badge/badge.stories.ts [hunk 4]
```diff
@@ -0,0 +1,220 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.stories.ts:1]: Do not disable eslint.

File: packages/components/src/components/badge/badge.test.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+/* eslint-disable @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any, @typescript-eslint/no-shadow, @typescript-eslint/no-unsafe-member-access  */
```
[the remark this answers is not in the document]
Comment (mariohamann) [reply packages/components/src/components/badge/badge.test.ts:1]: Do not disable eslint.

File: packages/components/src/components/badge/badge.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  @property({ type: Boolean, reflect: true }) inverted = false;
+
+  /** The badge's size. */
+  @property({ reflect: true }) size: 'lg' | 'md' | 'sm' = 'lg';
+
+  /** The badge's theme variant. */
+  @property({ reflect: true }) variant: 'default' | 'success' | 'error' = 'default';
+
+  /** Toggles the overflow indicator. */
+  @property({ type: Boolean, reflect: true }) overflowing = false;
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.ts:32]: This shouldn't exist.

**Comment by mariohamann** [comment]
@MarcMatthiae Always add designers to the PR, as they have to approve, too.

@solid-design-system/design

You can still access Storybook here:
CleanShot 2023-08-10 at 15 42 51@2x

**Comment by coraliefeil** [comment]
> CleanShot 2023-08-10 at 15 28 32@2x
> I'm not sure if this behaviour is correct. :)
> 
> I think it should be right-aligned, but maybe the designers have another opinion on that?

Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.

**Comment by coraliefeil** [comment]
> @MarcMatthiae Always add designers to the PR, as they have to approve, too.
> 
> @solid-design-system/design
> 
> You can still access Storybook here: CleanShot 2023-08-10 at 15 42 51@2x

Storybook publish link is missing ... is there another link?

**Comment by mariohamann** [comment]
> > @MarcMatthiae Always add designers to the PR, as they have to approve, too.
> > @solid-design-system/design
> > You can still access Storybook here: CleanShot 2023-08-10 at 15 42 51@2x
> 
> Storybook publish link is missing ... is there another link?

You have click on "Details", sorry.

**Comment by MarcMatthiae** [comment]
> > CleanShot 2023-08-10 at 15 28 32@2x
> > I'm not sure if this behaviour is correct. :)
> > I think it should be right-aligned, but maybe the designers have another opinion on that?
> 
> Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.

> > CleanShot 2023-08-10 at 15 28 32@2x
> > I'm not sure if this behaviour is correct. :)
> > I think it should be right-aligned, but maybe the designers have another opinion on that?
> 
> Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.

@mariohamann @coraliefeil Are you sure? Shoelace does it like I did.

**Comment by MarcMatthiae** [comment]
> > > CleanShot 2023-08-10 at 15 28 32@2x
> > > I'm not sure if this behaviour is correct. :)
> > > I think it should be right-aligned, but maybe the designers have another opinion on that?
> > 
> > 
> > Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.
> 
> > > CleanShot 2023-08-10 at 15 28 32@2x
> > > I'm not sure if this behaviour is correct. :)
> > > I think it should be right-aligned, but maybe the designers have another opinion on that?
> > 
> > 
> > Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.
> 
> @mariohamann @coraliefeil Are you sure? Shoelace does it like I did.

image

**Comment by mariohamann** [comment]
> > @mariohamann @coraliefeil Are you sure? Shoelace does it like I did.

Well Shoelace is our base, but regarding design we change stuff. Do you see problems in implementing it?

**Comment by mariohamann** [comment]
CleanShot 2023-08-11 at 16 26 36@2x

The numbers are not vertically centered here.

File: packages/components/src/components/badge/badge.test.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      expect(el.shadowRoot!.querySelector('[part~="overflow-indicator"]')).have.class('hidden');
+      expect(el.querySelector(`*[slot='overflow-indicator']`)).to.not.exist;
+    });
+
+    it('should default to default styling, with the primary color', async () => {
+      el = await fixture<SdBadge>(html` <sd-badge>Badge</sd-badge> `);
+      const part = el.shadowRoot!.querySelector('[part~="base"]')!;
+      expect(part.classList.value.trim()).to.contain(baseClasses);
+      expect(part.classList.value.trim()).to.contain(variants[0].tailwindClasses);
+    });
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.test.ts:65]: As we have visual testing (somethign Shoelace doesn't have!) I would say you can omit all tests that check for specific TailwindClasses.

In that case you're more testing if cx works, but that's not part of our job.

File: packages/components/src/components/badge/badge.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    css`
+      :host {
+        display: inline-flex;
+        align-items: center;
+        justify-content: center;
+      }
+
+      slot[name='overflow-indicator']::slotted(*) {
+        position: relative;
+        top: calc(((-1em / 1.5) + 1ex) / 2);
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.ts:84]: Please provide some explenation why you came up with this calculation.

**Comment by MarcMatthiae** [comment]
> CleanShot 2023-08-11 at 16 26 36@2x
> 
> The numbers are not vertically centered here.

@mariohamann @coraliefeil Can we change the font size for the medium badge to 11? With the current font size of 10 the text is not vertically centered inside the badge due to the font.

**Comment by MarcMatthiae** [comment]
> > > > CleanShot 2023-08-10 at 15 28 32@2x
> > > > I'm not sure if this behaviour is correct. :)
> > > > I think it should be right-aligned, but maybe the designers have another opinion on that?
> > > 
> > > 
> > > Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.
> > 
> > 
> > > > CleanShot 2023-08-10 at 15 28 32@2x
> > > > I'm not sure if this behaviour is correct. :)
> > > > I think it should be right-aligned, but maybe the designers have another opinion on that?
> > > 
> > > 
> > > Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.
> > 
> > 
> > @mariohamann @coraliefeil Are you sure? Shoelace does it like I did.
> 
> image

Actually, this needs to be done in the button component, right?

**Comment by mariohamann** [comment]
> > > > > CleanShot 2023-08-10 at 15 28 32@2x
> > > > > I'm not sure if this behaviour is correct. :)
> > > > > I think it should be right-aligned, but maybe the designers have another opinion on that?
> > > > 
> > > > 
> > > > Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.
> > > 
> > > 
> > > > > CleanShot 2023-08-10 at 15 28 32@2x
> > > > > I'm not sure if this behaviour is correct. :)
> > > > > I think it should be right-aligned, but maybe the designers have another opinion on that?
> > > > 
> > > > 
> > > > Yes, thank you for pointing that out @mariohamann. 1 digit is centered in the element corner. More digits - width increases to the left / badge is right-aligned.
> > > 
> > > 
> > > @mariohamann @coraliefeil Are you sure? Shoelace does it like I did.
> > 
> > 
> > image
> 
> Actually, this needs to be done in the button component, right?

Yes. If you want, you can leave it out completely at this point and we wait until design provides the sample, too.

**Comment by MarcMatthiae** [comment]
Pipeline failing because of color accessibility problems when using token bg-success. Waiting for token update.

File: packages/components/src/components/badge/badge.stories.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+/**
+ * Use the `default` slot to add content to the badge.
+ * Use the `overflow-indicator` slot to add an overflow indicator.
+ * *
+ * <b>Hints</b>
+ * - If you add icons to one of the slots, please make sure to account for accessibility by providing an alt-text.
+ * - If <b>not</b> passed an element to the overflow-indicator slot, the badge will <b>not</b> add an overflow-indicator per default.
+ * - Content in the overflow-indicator slot is not centered vertically per default, as we do not know what will end up in the slot.
+ */
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.stories.ts:140]: Could please write this as markdown instead of HTML?

**Review by mariohamann — APPROVED** [review]

**Comment by mariohamann** [comment]
@solid-design-system/design Did you already give your approval?

**Comment by coraliefeil** [comment]
> @solid-design-system/design Did you already give your approval?

Review done. Just 2 tiny things. Then we are good to go. Thank you all <3

File: packages/components/src/components/badge/badge.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+          }[this.variant],
+          {
+            /* size and fonts*/
+            sm: 'h-2 min-w-[8px] text-[10px]',
+            md: 'h-4 px-[4px] min-w-[16px] text-[10px]',
+            lg: 'h-5 px-[5px] min-w-[20px] text-[12px]'
+          }[this.size]
+        )}
+      >
+        <span part="content" class=${cx(this.size === 'sm' && `hidden`)}>
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.ts:50]:
```suggestion
        <span part="content" class=${cx(this.size === 'sm' && 'hidden')}>
```

File: packages/components/src/components/badge/badge.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        tabindex="0"
+        role="status"
+        part="base"
+        class=${cx(
+          `inline-flex items-center justify-center gap-x-[1px] text-center leading-none whitespace-nowrap border rounded-full select-none cursor-[inherit]`,
+          {
+            /* variants */
+            default: !this.inverted ? `text-white bg-primary-500 border-white` : `text-primary bg-white border-primary`,
+            success: !this.inverted ? `text-white bg-success border-white` : `text-white bg-success border-primary`,
+            error: !this.inverted ? `text-white bg-error border-white` : `text-white bg-error border-primary`
```
Comment (mariohamann) [inline packages/components/src/components/badge/badge.ts:40]:
```suggestion
          'inline-flex items-center justify-center gap-x-[1px] text-center leading-none whitespace-nowrap border rounded-full select-none cursor-[inherit]',
          {
            /* variants */
            default: !this.inverted ? 'text-white bg-primary-500 border-white' : 'text-primary bg-white border-primary',
            success: !this.inverted ? 'text-white bg-success border-white' : 'text-white bg-success border-primary',
            error: !this.inverted ? 'text-white bg-error border-white' : 'text-white bg-error border-primary'
```

**Pull request merged**